compiler/noirc_frontend/src/elaborator/statements.rs

```diff
@@ -271,11 +271,25 @@ impl<'context> Elaborator<'context> {
 
     pub(super) fn elaborate_loop(
         &mut self,
-        _block: Expression,
+        block: Expression,
         span: noirc_errors::Span,
     ) -> (HirStatement, Type) {
-        self.push_err(ResolverError::LoopNotYetSupported { span });
-        (HirStatement::Error, Type::Unit)
+        let in_constrained_function = self.in_constrained_function();
+        if in_constrained_function {
+            self.push_err(ResolverError::LoopInConstrainedFn { span });
+        }
+
+        self.nested_loops += 1;
+        self.push_scope();
+
+        let (block, _block_type) = self.elaborate_expression(block);
+
+        self.pop_scope();
+        self.nested_loops -= 1;
+
+        let statement = HirStatement::Loop(block);
+
+        (statement, Type::Unit)
```

Comment:
We should have loops return the value that `break` was used with if possible. We can save this for a later PR though.

Reply:
Good point. I think `break` doesn't allow expressions yet. And if we make this change, maybe `for` should also allow that.

Reply:
`for` can't also allow that since like in rust it may exit without a break, so there'd be no value to return. We could at most return a `Option<T>` but that gets awkward - we should just keep following Rust here.